File: apps/api/src/repository/GroupRepository.ts
import dbClient from "../dbconfig";

class GroupRepository {

    static async getAllGroups() {
        const res = await dbClient.query('SELECT * FROM schedule.groups');
        return res.rows;
    }

    static async getGroupById(group_id: string) {
        const res = await dbClient.query('SELECT * FROM schedule.groups where group_id = $1', [group_id]);
        return res.rows[0];
    }

    static async createGroup(group_id: string, group_name: string) {
        const res = await dbClient.query(
            'INSERT INTO schedule.groups (group_id, group_name) VALUES ($1, $2) RETURNING *',
            [group_id, group_name]
        );
        return res.rows[0];
    }

    static async updateGroup(group_id: string, group_name: string) {
        const res = await dbClient.query(
            'UPDATE schedule.groups SET group_name = $1 WHERE group_id = $2 RETURNING *',
            [group_name, group_id]
        );
        return res.rows[0];
    }

    static async deleteGroup(group_id: string) {
        await dbClient.query(
            'DELETE FROM schedule.groups where group_id = $1',
            [group_id]
        );
    }
}

export default GroupRepository;

File: apps/cs-api/Repositories/GroupRepository.cs
using cs_api.Config;
using cs_api.Models;
using Npgsql;

namespace cs_api.Repositories;

public class GroupRepository : IGroupRepository
{
    private readonly NpgsqlDataSource _dataSource;

    public GroupRepository(DbConfig dbConfig)
    {
        _dataSource = dbConfig.DataSource;
    }

    public async Task<IEnumerable<Group>> GetAllGroups()
    {
        await using var conn = await _dataSource.OpenConnectionAsync();
        await using var cmd = new NpgsqlCommand(
            "SELECT group_id, group_name FROM schedule.groups", conn);
        await using var reader = await cmd.ExecuteReaderAsync();

        var groups = new List<Group>();
        while (await reader.ReadAsync())
        {
            groups.Add(MapGroup(reader));
        }
        return groups;
    }

    public async Task<Group?> GetGroupById(string groupId)
    {
        await using var conn = await _dataSource.OpenConnectionAsync();
        await using var cmd = new NpgsqlCommand(
            "SELECT group_id, group_name FROM schedule.groups WHERE group_id = @id", conn);
        cmd.Parameters.AddWithValue("id", Guid.Parse(groupId));
        await using var reader = await cmd.ExecuteReaderAsync();

        return await reader.ReadAsync() ? MapGroup(reader) : null;
    }

    public async Task<Group> CreateGroup(string groupId, string groupName)
    {
        await using var conn = await _dataSource.OpenConnectionAsync();
        await using var cmd = new NpgsqlCommand(
            "INSERT INTO schedule.groups (group_id, group_name) VALUES (@id, @name) RETURNING group_id, group_name", conn);
        cmd.Parameters.AddWithValue("id", Guid.Parse(groupId));
        cmd.Parameters.AddWithValue("name", groupName);
        await using var reader = await cmd.ExecuteReaderAsync();

        await reader.ReadAsync();
        return MapGroup(reader);
    }

    public async Task<Group?> UpdateGroup(string groupId, string? groupName)
    {
        if (groupName == null) return null;

        await using var conn = await _dataSource.OpenConnectionAsync();
        await using var cmd = new NpgsqlCommand(
            "UPDATE schedule.groups SET group_name = @name WHERE group_id = @id RETURNING group_id, group_name", conn);
        cmd.Parameters.AddWithValue("name", groupName);
        cmd.Parameters.AddWithValue("id", Guid.Parse(groupId));
        await using var reader = await cmd.ExecuteReaderAsync();

        return await reader.ReadAsync() ? MapGroup(reader) : null;
    }

    public async Task DeleteGroup(string groupId)
    {
        await using var conn = await _dataSource.OpenConnectionAsync();
        await using var cmd = new NpgsqlCommand(
            "DELETE FROM schedule.groups WHERE group_id = @id", conn);
        cmd.Parameters.AddWithValue("id", Guid.Parse(groupId));
        await cmd.ExecuteNonQueryAsync();
    }

    private static Group MapGroup(NpgsqlDataReader reader) => new()
    {
        GroupId = reader.GetGuid(0).ToString(),
        GroupName = reader.GetString(1)
    };
}